xds/src/main/java/io/grpc/xds/EdsLoadBalancer.java

```diff
@@ -0,0 +1,715 @@
+import io.grpc.ConnectivityState;
+import io.grpc.EquivalentAddressGroup;
+import io.grpc.LoadBalancer;
+import io.grpc.LoadBalancerProvider;
+import io.grpc.LoadBalancerRegistry;
+import io.grpc.Status;
+import io.grpc.SynchronizationContext.ScheduledHandle;
+import io.grpc.util.ForwardingLoadBalancerHelper;
+import io.grpc.xds.ClientLoadCounter.LoadRecordingSubchannelPicker;
+import io.grpc.xds.ClientLoadCounter.MetricsObservingSubchannelPicker;
+import io.grpc.xds.ClientLoadCounter.MetricsRecordingListener;
+import io.grpc.xds.EnvoyProtoData.DropOverload;
+import io.grpc.xds.EnvoyProtoData.LbEndpoint;
+import io.grpc.xds.EnvoyProtoData.Locality;
+import io.grpc.xds.EnvoyProtoData.LocalityLbEndpoints;
+import io.grpc.xds.InterLocalityPicker.WeightedChildPicker;
+import io.grpc.xds.OrcaOobUtil.OrcaReportingConfig;
+import io.grpc.xds.OrcaOobUtil.OrcaReportingHelperWrapper;
+import io.grpc.xds.XdsClient.EndpointUpdate;
+import io.grpc.xds.XdsClient.EndpointWatcher;
+import io.grpc.xds.XdsSubchannelPickers.ErrorPicker;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+
+/**
+ * Load balancer for experimental_eds LB policy.
+ */
+final class EdsLoadBalancer extends LoadBalancer {
+
+  private final Helper helper;
+  private final XdsClient xdsClient;
+  private final ChannelLogger channelLogger;
+
+  private final EndpointWatcherImpl.PickerFactory pickerFactory;
+  private final LoadBalancerRegistry lbRegistry;
+  private final ThreadSafeRandom random;
+  private final LoadStatsStore loadStatsStore;
+  private final OrcaPerRequestUtil orcaPerRequestUtil;
+  private final OrcaOobUtil orcaOobUtil;
+
+
+  @Nullable
+  private String edsServiceName;
+
+  @CheckForNull
+  private EndpointWatcherImpl endpointWatcher;
+
+  // TODO(zdapeng): Pass in a nullable LoadStatsStore for LRS.
+  EdsLoadBalancer(Helper helper, XdsClient xdsClient) {
+    this(
+        helper, xdsClient, EndpointWatcherImpl.pickerFactoryImpl,
+        LoadBalancerRegistry.getDefaultRegistry(), ThreadSafeRandom.ThreadSafeRandomImpl.instance,
+        new LoadStatsStoreImpl(), OrcaPerRequestUtil.getInstance(), OrcaOobUtil.getInstance());
+  }
+
+  @VisibleForTesting
+  EdsLoadBalancer(
+      Helper helper,
+      XdsClient xdsClient,
+      EndpointWatcherImpl.PickerFactory pickerFactory,
+      LoadBalancerRegistry lbRegistry,
+      ThreadSafeRandom random,
+      LoadStatsStore loadStatsStore,
+      OrcaPerRequestUtil orcaPerRequestUtil,
+      OrcaOobUtil orcaOobUtil) {
+    this.helper = checkNotNull(helper, "helper");
+    this.xdsClient = checkNotNull(xdsClient, "xdsClient");
+    this.channelLogger = helper.getChannelLogger();
+    // TODO(zdapeng): Allow it be null when enableLrs from ClusterUpdate is false.
```

Comment:
Loads may still need to be recorded even if `enableLrs` is false, i.e., start load reporting later. But not 100% sure about this requirement, will need to confirm with design.